# Segmenting and Clustering Neighborhoods in the city of Toronto, Canada

In [2]:
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
from urllib import request
import urllib

In [3]:
wikipedia_url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

html = urllib.request.urlopen(wikipedia_url).read()
soup = BeautifulSoup(html, "html.parser")

### Steps
* Read the table in Wikipedia
* Go through one table row at a time
* Add each table row to a dictionary, which can then be used to create the DataFrame

In [4]:
results = []
table = soup.find("table",{"class":"wikitable sortable"})
trs = table.findAll('tr')
trs = [tr for tr in trs if len(tr.find_all('th'))<1 ]
for tr in trs:
    tds = tr.find_all('td')
    #     PostalCode, Borough, and Neighborhood
    d = {
        'PostalCode': str(tds[0].text.strip()),
        'Borough': str(tds[1].text.strip()),
        'Neighborhood': str(tds[2].text.strip())
    }
    results.append(d)


### Create the DataFrame

In [5]:
df = pd.DataFrame(results)
df.head()

Borough      Neighborhood PostalCode
0      Not assigned      Not assigned        M1A
1      Not assigned      Not assigned        M2A
2        North York         Parkwoods        M3A
3        North York  Victoria Village        M4A
4  Downtown Toronto      Harbourfront        M5A

### Ignore cells with a borough that is Not assigned.

In [6]:
df = df[df.Borough != 'Not assigned']

### Combine Neighbourhoods belonging to same PostalCode and Borough

In [7]:
df = df.groupby(['PostalCode', 'Borough']).agg(','.join).reset_index()
df[df.PostalCode == 'M9V']

PostalCode    Borough                                       Neighborhood
101        M9V  Etobicoke  Albion Gardens,Beaumond Heights,Humbergate,Jam...

### If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [8]:
df['Neighborhood'] = np.where(df.Neighborhood == 'Not assigned', df.Borough, df.Neighborhood)

In [9]:
df[df.PostalCode == 'M7A']

PostalCode       Borough  Neighborhood
85        M7A  Queen's Park  Queen's Park

### Total Number of Rows

In [10]:
df.shape

(103, 3)

## *** Second Question ***

#### I created a separate notebook as the instructions were a bit ambiguous

In [11]:
lat_lng_file = 'https://cocl.us/Geospatial_data'
lat_lng = pd.read_csv(lat_lng_file)
lat_lng.head(3)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711

### Merge the Lat Lng Dataframe with the Locations DF

In [12]:
df = df.merge(lat_lng, left_on='PostalCode', right_on = 'Postal Code', how='left')

In [13]:
df.head()

PostalCode      Borough                          Neighborhood Postal Code  \
0        M1B  Scarborough                         Rouge,Malvern         M1B   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union         M1C   
2        M1E  Scarborough       Guildwood,Morningside,West Hill         M1E   
3        M1G  Scarborough                                Woburn         M1G   
4        M1H  Scarborough                             Cedarbrae         M1H   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

#### We can drop one extra Postal Code column

In [14]:
df.drop('Postal Code', axis=1, inplace=True)

### Make sure we have not increased the number of rows. It should still be 103 rows and 5 columns

In [15]:
df.shape

(103, 5)

In [16]:
df.head(3)

PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711

# **** Third Question ****

### Explore & Cluster neighborhoods in Toronto

In [23]:
print("The boroughs in Toronto are : %s"%df[df.Borough.str.contains('Toronto')].Borough.unique().tolist())

The boroughs in Toronto are : ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']


In [24]:
toronto_df = df[df.Borough.str.contains('Toronto')]
toronto_df.shape

(38, 5)

In [45]:
import folium
import requests
from geopy.geocoders import Nominatim 
from pandas.io.json import json_normalize

#### Get Toronto Coordinates

In [27]:
address = 'Toronto, Canada'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [493]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=4,
        popup=label,
        color='purple',
        fill=True,
        fill_color='#2186cc',
        fill_opacity=0.3,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [494]:
# @hidden cell
CLIENT_ID = 'DFELEIN3LFIR3CVHUFUOCAP1CT4THHW3Y0YUKIGPKBOIIJQZ' 
CLIENT_SECRET = 'LV0LAP0DY0MYA4QGRCF5EFD1B2DGTDPE10DFWZESBBP3KKE1' 
VERSION = '20180605' # Foursquare API version


In [39]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [40]:
toronto_df.head()

PostalCode          Borough                   Neighborhood   Latitude  \
37        M4E     East Toronto                    The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
42        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
43        M4M     East Toronto                Studio District  43.659526   
44        M4N  Central Toronto                  Lawrence Park  43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

### For each Postal Code, find the top 100 venues using FourSquare APIs, restricted to a radius of 1 Km

In [175]:
postcodes = toronto_df.PostalCode.unique().tolist()
print("We have %s postal codes to search in Toronto"%len(postcodes))

We have 38 postal codes to search in Toronto


In [176]:
radius = 1000
limit = 100
venues_list=[]
for postcode in postcodes:
    lat = toronto_df[toronto_df.PostalCode == postcode].iloc[0]['Latitude']
    lng = toronto_df[toronto_df.PostalCode == postcode].iloc[0]['Longitude']
    url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        CLIENT_ID, 
        CLIENT_SECRET, 
        VERSION, 
        lat, 
        lng, 
        radius, 
        limit)

    print("Seaching for postal code = %s"%postcode)
    results = requests.get(url).json()
    venues = results['response']['groups'][0]['items']
    venues_list.append([(
            postcode, 
            toronto_df[toronto_df.PostalCode == postcode].iloc[0]['Borough'],
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in venues])
    

Seaching for postal code = M4E
Seaching for postal code = M4K
Seaching for postal code = M4L
Seaching for postal code = M4M
Seaching for postal code = M4N
Seaching for postal code = M4P
Seaching for postal code = M4R
Seaching for postal code = M4S
Seaching for postal code = M4T
Seaching for postal code = M4V
Seaching for postal code = M4W
Seaching for postal code = M4X
Seaching for postal code = M4Y
Seaching for postal code = M5A
Seaching for postal code = M5B
Seaching for postal code = M5C
Seaching for postal code = M5E
Seaching for postal code = M5G
Seaching for postal code = M5H
Seaching for postal code = M5J
Seaching for postal code = M5K
Seaching for postal code = M5L
Seaching for postal code = M5N
Seaching for postal code = M5P
Seaching for postal code = M5R
Seaching for postal code = M5S
Seaching for postal code = M5T
Seaching for postal code = M5V
Seaching for postal code = M5W
Seaching for postal code = M5X
Seaching for postal code = M6G
Seaching for postal code = M6H
Seaching

#### Build a Dataframe containing all the venues

In [495]:
nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
nearby_venues.columns = ['PostalCode', 'Borough',
              'PostalCode Latitude', 
              'PostalCode Longitude', 
              'Venue', 
              'Venue Latitude', 
              'Venue Longitude', 
              'Venue Category']


In [496]:
print("Found %s Categories"%len(nearby_venues['Venue Category'].unique().tolist()))

Found 276 Categories


### That's a lot of categories !!!
### Let's look at the Top 15 Categories....and see if we can combine some of them.

In [497]:
nearby_venues.groupby(['Venue Category']).count().sort_values('PostalCode', ascending=False).reset_index().head(15)

Venue Category  PostalCode  Borough  PostalCode Latitude  \
0                     Coffee Shop         234      234                  234   
1                            Café         170      170                  170   
2              Italian Restaurant          91       91                   91   
3                      Restaurant          78       78                   78   
4                          Bakery          77       77                   77   
5                            Park          75       75                   75   
6                     Pizza Place          62       62                   62   
7                             Bar          62       62                   62   
8                           Hotel          62       62                   62   
9                Sushi Restaurant          56       56                   56   
10            Japanese Restaurant          53       53                   53   
11                            Gym          50       50                   50   
12                      Gastropub          48       48                   48   
13                            Pub          46       46                   46   
14  Vegetarian / Vegan Restaurant          43       43                   43   

    PostalCode Longitude  Venue  Venue Latitude  Venue Longitude  
0                    234    234             234              234  
1                    170    170             170              170  
2                     91     91              91               91  
3                     78     78              78               78  
4                     77     77              77               77  
5                     75     75              75               75  
6                     62     62              62               62  
7                     62     62              62               62  
8                     62     62              62               62  
9                     56     56              56               56  
10                    53     53              53               53  
11                    50     50              50               50  
12                    48     48              48               48  
13                    46     46              46               46  
14                    43     43              43               43

#### Seems like we have plenty of Restaurants (Indian, Chineese, Turkish, Italian, Greek, Pizza, Steakhouse, etc.) . Let's group them under a generic "Restaurant" category

In [498]:
mask = nearby_venues['Venue Category'].str.contains('Restaurant')
nearby_venues.loc[mask, 'Venue Category'] = 'Restaurant'
nearby_venues['Venue Category'] = np.where(nearby_venues['Venue Category'].isin(['Diner', 'Burger Joint', 'Burrito Place', 'Steakhouse', 'Pizza Place']), 'Restaurant', nearby_venues['Venue Category'])

#### Merge Coffee Shops, Cafe, Bakery, Breakfast spot, Tea Room into a single category called "Cafe"

In [499]:
nearby_venues['Venue Category'] = np.where(nearby_venues['Venue Category'].isin(['Café', 'Sandwich Place', 'Tea Room', 'Coffee Shop', 'Bakery', 'Breakfast Spot', 'Ice Cream Shop']), 'Café', nearby_venues['Venue Category'])

#### Merge Pub, Bar, GastroPub into a single category

In [500]:
nearby_venues['Venue Category'] = np.where(nearby_venues['Venue Category'].isin(['Pub', 'Gastropub', 'Bar', 'Beer Bar', 'Cocktail Bar', 'Brewery']), 'Bar', nearby_venues['Venue Category'])

In [501]:
print("Found %s Categories after grouping restaurants"%len(nearby_venues['Venue Category'].unique().tolist()))

Found 207 Categories after grouping restaurants


In [502]:
nearby_venues.groupby(['Venue Category']).count().sort_values('PostalCode', ascending=False).reset_index().head(10)

Venue Category  PostalCode  Borough  PostalCode Latitude  \
0     Restaurant         944      944                  944   
1           Café         612      612                  612   
2            Bar         229      229                  229   
3           Park          75       75                   75   
4          Hotel          62       62                   62   
5            Gym          50       50                   50   
6  Grocery Store          32       32                   32   
7        Theater          32       32                   32   
8      Bookstore          30       30                   30   
9  Deli / Bodega          24       24                   24   

   PostalCode Longitude  Venue  Venue Latitude  Venue Longitude  
0                   944    944             944              944  
1                   612    612             612              612  
2                   229    229             229              229  
3                    75     75              75               75  
4                    62     62              62               62  
5                    50     50              50               50  
6                    32     32              32               32  
7                    32     32              32               32  
8                    30     30              30               30  
9                    24     24              24               24

In [503]:
nearby_venues.shape

(3089, 8)

### Analyse Each Postal Code Neighborhood
### Looks like Toronto is predominantly dominated the first 3 categories - Restaurants, Bars and Cafes

In [505]:
# one hot encoding
toronto_onehot = pd.get_dummies(nearby_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot['PostalCode'] = nearby_venues['PostalCode'] 
toronto_onehot.head()

Accessories Store  Airport  Airport Lounge  Amphitheater  Animal Shelter  \
0                  0        0               0             0               0   
1                  0        0               0             0               0   
2                  0        0               0             0               0   
3                  0        0               0             0               0   
4                  0        0               0             0               0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

      ...      University  Video Game Store  Video Store  Whisky Bar  \
0     ...               0                 0            0           0   
1     ...               0                 0            0           0   
2     ...               0                 0            0           0   
3     ...               0                 0            0           0   
4     ...               0                 0            0           0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  Zoo  PostalCode  
0         0            0              0            0    0         M4E  
1         0            0              0            0    0         M4E  
2         0            0              0            0    0         M4E  
3         0            0              0            0    0         M4E  
4         0            0              0            0    0         M4E  

[5 rows x 208 columns]

### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [506]:
toronto_grouped = toronto_onehot.groupby('PostalCode').mean().reset_index()
toronto_grouped.head()

PostalCode  Accessories Store  Airport  Airport Lounge  Amphitheater  \
0        M4E                0.0      0.0             0.0           0.0   
1        M4K                0.0      0.0             0.0           0.0   
2        M4L                0.0      0.0             0.0           0.0   
3        M4M                0.0      0.0             0.0           0.0   
4        M4N                0.0      0.0             0.0           0.0   

   Animal Shelter  Antique Shop  Aquarium  Art Gallery  Art Museum ...   \
0             0.0          0.00       0.0     0.000000         0.0 ...    
1             0.0          0.00       0.0     0.000000         0.0 ...    
2             0.0          0.00       0.0     0.012821         0.0 ...    
3             0.0          0.01       0.0     0.000000         0.0 ...    
4             0.0          0.00       0.0     0.000000         0.0 ...    

   Train Station  University  Video Game Store  Video Store  Whisky Bar  \
0            0.0         0.0               0.0          0.0         0.0   
1            0.0         0.0               0.0          0.0         0.0   
2            0.0         0.0               0.0          0.0         0.0   
3            0.0         0.0               0.0          0.0         0.0   
4            0.0         0.0               0.0          0.0         0.0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  Zoo  
0       0.0          0.0       0.012048     0.000000  0.0  
1       0.0          0.0       0.000000     0.020000  0.0  
2       0.0          0.0       0.012821     0.012821  0.0  
3       0.0          0.0       0.000000     0.010000  0.0  
4       0.0          0.0       0.000000     0.000000  0.0  

[5 rows x 208 columns]

### For each postal code we find the Top 3 categories

In [507]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]

In [508]:
num_top_venues = 5
indicators = ['st', 'nd', 'rd']
# create columns according to number of top venues
columns = ['PostalCode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_grouped['PostalCode']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

PostalCode 1st Most Common Venue 2nd Most Common Venue  \
0        M4E            Restaurant                  Café   
1        M4K            Restaurant                  Café   
2        M4L            Restaurant                  Café   
3        M4M            Restaurant                  Café   
4        M4N                  Café             Bookstore   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
0                   Bar                 Beach                  Park  
1                   Bar           Yoga Studio                  Bank  
2                   Bar                 Beach                   Gym  
3                   Bar  Gym / Fitness Center                  Park  
4          College Quad  Gym / Fitness Center           College Gym

### Start Clustering Neighborhoods into 3 clusters

In [509]:
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn import metrics

In [510]:
kclusters = 3
print("Number of Clusters = %s"%kclusters)

Number of Clusters = 3


In [511]:
toronto_clusters = toronto_grouped.drop('PostalCode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=5).fit(toronto_clusters)


In [512]:
# ADD CLUSTERED LABELS AND CREATE A NEW MERGED DF
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
toronto_df_with_clusters = toronto_df
toronto_df_with_clusters = toronto_df_with_clusters.join(neighborhoods_venues_sorted.set_index('PostalCode'), on='PostalCode')
toronto_df_with_clusters.head()

PostalCode          Borough                   Neighborhood   Latitude  \
37        M4E     East Toronto                    The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
42        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
43        M4M     East Toronto                Studio District  43.659526   
44        M4N  Central Toronto                  Lawrence Park  43.728020   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37 -79.293031               0            Restaurant                  Café   
41 -79.352188               1            Restaurant                  Café   
42 -79.315572               1            Restaurant                  Café   
43 -79.340923               0            Restaurant                  Café   
44 -79.388790               2                  Café             Bookstore   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
37                   Bar                 Beach                  Park  
41                   Bar           Yoga Studio                  Bank  
42                   Bar                 Beach                   Gym  
43                   Bar  Gym / Fitness Center                  Park  
44          College Quad  Gym / Fitness Center           College Gym

### Visualize the Clusters

In [513]:
# create map
address = 'Toronto, Canada'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
# rainbow = [colors.rgb2hex(i) for i in colors_array]
rainbow = ['blue', 'red', 'green', 'yellow', 'orange', 'purple', 'grey']
# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_df_with_clusters['Latitude'], toronto_df_with_clusters['Longitude'], toronto_df_with_clusters['PostalCode'], toronto_df_with_clusters['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Analyse Clusters

## Cluster 1

In [514]:
toronto_df_with_clusters[toronto_df_with_clusters['Cluster Labels'] == 0]

PostalCode           Borough  \
37        M4E      East Toronto   
43        M4M      East Toronto   
53        M5A  Downtown Toronto   
55        M5C  Downtown Toronto   
56        M5E  Downtown Toronto   
59        M5J  Downtown Toronto   
61        M5L  Downtown Toronto   
63        M5N   Central Toronto   
64        M5P   Central Toronto   
69        M5W  Downtown Toronto   
76        M6H      West Toronto   
77        M6J      West Toronto   
78        M6K      West Toronto   
82        M6P      West Toronto   
83        M6R      West Toronto   
84        M6S      West Toronto   

                                       Neighborhood   Latitude  Longitude  \
37                                      The Beaches  43.676357 -79.293031   
43                                  Studio District  43.659526 -79.340923   
53                         Harbourfront,Regent Park  43.654260 -79.360636   
55                                   St. James Town  43.651494 -79.375418   
56                                      Berczy Park  43.644771 -79.373306   
59  Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752   
61                    Commerce Court,Victoria Hotel  43.648198 -79.379817   
63                                         Roselawn  43.711695 -79.416936   
64               Forest Hill North,Forest Hill West  43.696948 -79.411307   
69                  Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
76                      Dovercourt Village,Dufferin  43.669005 -79.442259   
77                          Little Portugal,Trinity  43.647927 -79.419750   
78       Brockton,Exhibition Place,Parkdale Village  43.636847 -79.428191   
82                     High Park,The Junction South  43.661608 -79.464763   
83                            Parkdale,Roncesvalles  43.648960 -79.456325   
84                                Runnymede,Swansea  43.651571 -79.484450   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37               0            Restaurant                  Café   
43               0            Restaurant                  Café   
53               0            Restaurant                  Café   
55               0            Restaurant                  Café   
56               0            Restaurant                  Café   
59               0                  Café            Restaurant   
61               0            Restaurant                  Café   
63               0            Restaurant                  Café   
64               0            Restaurant                  Café   
69               0            Restaurant                  Café   
76               0                  Café            Restaurant   
77               0            Restaurant                  Café   
78               0            Restaurant                  Café   
82               0            Restaurant                  Café   
83               0            Restaurant                  Café   
84               0            Restaurant                  Café   

   3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  
37                   Bar                   Beach                  Park  
43                   Bar    Gym / Fitness Center                  Park  
53                   Bar                 Theater                  Park  
55                   Bar                   Hotel        Cosmetics Shop  
56                   Bar                   Hotel                  Park  
59                   Bar                   Hotel              Aquarium  
61                   Bar                   Hotel          Concert Hall  
63                  Bank                Pharmacy          Dance Studio  
64  Gym / Fitness Center            Skating Rink          Liquor Store  
69                   Bar                   Hotel                   Gym  
76                  Park                     Bar              Pharmacy  
77                   Bar               Gift Shop               Theater  
78                   Bar 

#### Cluster 1 looks like its locations with Restaurants and Cafes as the 1st and 2nd Most common venues

## Cluster 2

In [515]:
toronto_df_with_clusters[toronto_df_with_clusters['Cluster Labels'] == 1]

PostalCode           Borough  \
41        M4K      East Toronto   
42        M4L      East Toronto   
45        M4P   Central Toronto   
46        M4R   Central Toronto   
47        M4S   Central Toronto   
48        M4T   Central Toronto   
49        M4V   Central Toronto   
51        M4X  Downtown Toronto   
52        M4Y  Downtown Toronto   
54        M5B  Downtown Toronto   
57        M5G  Downtown Toronto   
58        M5H  Downtown Toronto   
60        M5K  Downtown Toronto   
65        M5R   Central Toronto   
66        M5S  Downtown Toronto   
67        M5T  Downtown Toronto   
70        M5X  Downtown Toronto   
75        M6G  Downtown Toronto   
87        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
41                        The Danforth West,Riverdale  43.679557 -79.352188   
42                      The Beaches West,India Bazaar  43.668999 -79.315572   
45                                   Davisville North  43.712751 -79.390197   
46                                 North Toronto West  43.715383 -79.405678   
47                                         Davisville  43.704324 -79.388790   
48                         Moore Park,Summerhill East  43.689574 -79.383160   
49  Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049   
51                         Cabbagetown,St. James Town  43.667967 -79.367675   
52                               Church and Wellesley  43.665860 -79.383160   
54                            Ryerson,Garden District  43.657162 -79.378937   
57                                 Central Bay Street  43.657952 -79.387383   
58                             Adelaide,King,Richmond  43.650571 -79.384568   
60            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576   
65                  The Annex,North Midtown,Yorkville  43.672710 -79.405678   
66                      Harbord,University of Toronto  43.662696 -79.400049   
67            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049   
70              First Canadian Place,Underground city  43.648429 -79.382280   
75                                           Christie  43.669542 -79.422564   
87  Business Reply Mail Processing Centre 969 Eastern  43.662744 -79.321558   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
41               1            Restaurant                  Café   
42               1            Restaurant                  Café   
45               1            Restaurant                  Café   
46               1            Restaurant                  Café   
47               1            Restaurant                  Café   
48               1            Restaurant                  Café   
49               1            Restaurant                  Café   
51               1            Restaurant                  Café   
52               1            Restaurant                  Café   
54               1            Restaurant                  Café   
57               1            Restaurant                  Café   
58               1            Restaurant                  Café   
60               1            Restaurant                  Café   
65               1            Restaurant                  Café   
66               1            Restaurant                  Café   
67               1            Restaurant                  Café   
70               1            Restaurant                  Café   
75               1            Restaurant                  Café   
87               1            Restaurant                  Café   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
41                   Bar           Yoga Studio                  Bank  
42                   Bar                 Beach                   Gym  
45          Dessert Shop                   Bar                   Gym  
46   Sporting Goods Shop          Skating Rink  Gym / Fitness Center  
47                   Bar                   Gy

#### Cluster 2 looks like its locations with Restaurants as the first common venue and Cafes as the 2nd Most common venues

## Cluster 3

In [516]:
toronto_df_with_clusters[toronto_df_with_clusters['Cluster Labels'] == 2]

PostalCode           Borough  \
44        M4N   Central Toronto   
50        M4W  Downtown Toronto   
68        M5V  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
44                                      Lawrence Park  43.728020 -79.388790   
50                                           Rosedale  43.679563 -79.377529   
68  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
44               2                  Café             Bookstore   
50               2                  Café         Grocery Store   
68               2                  Café       Harbor / Marina   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
44          College Quad  Gym / Fitness Center           College Gym  
50                  Park            Restaurant                  Bank  
68                  Park                Garden      Sculpture Garden

#### Cluster 3 looks like areas that are mixed with Cafes, Parks, and other facilities in Toronto.....Could these be more residential?